pkg/controller/engine.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package controller provides utilties for working with controllers.
+package controller
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Error strings
+const (
+	errCreateCache      = "cannot create new cache"
+	errCreateController = "cannot create new controller"
+	errCrashCache       = "cache error"
+	errCrashController  = "controller error"
+	errWatch            = "cannot setup watch"
+)
+
+// A NewCacheFn creates a new controller-runtime cache.
+type NewCacheFn func(cfg *rest.Config, o cache.Options) (cache.Cache, error)
+
+// A NewControllerFn creates a new controller-runtime controller.
+type NewControllerFn func(name string, m manager.Manager, o controller.Options) (controller.Controller, error)
+
+// The default new cache and new controller functions.
+var (
+	DefaultNewCacheFn      NewCacheFn      = cache.New
+	DefaultNewControllerFn NewControllerFn = controller.NewUnmanaged
+)
+
+// An Engine manages the lifecycles of controller-runtime controllers (and their
+// caches). The lifecycles of the controllers are not coupled to lifecycle of
+// the engine, nor to the lifecycle of the controller manager it uses.
+type Engine struct {
```

Comment:
Will this replace https://github.com/crossplane/crossplane/blob/fc494b11777dda20e2c1f70d9f627c390324f7e4/pkg/controller/apiextensions/composite/engine.go#L44?

Reply:
Yeah, that's the idea. I liked the `ControllerEngine` abstraction around the controller-runtime machinery, and it seems like we'll probably have at least 2-3 controllers that are managing groups of child controllers, so I wanted to clean it up a little and add it here.